### PR TITLE
Support for handling default back button in JS

### LINKED ISSUE #6935: Adding an id to a backbutton makes it invisible
## 🐛 Bug Report

Adding an id to a back button make it invisible even if `visible: true`

## To Reproduce

https://github.com/Andarius/react-native-navigation/tree/issue/topbarBackBtn

1. Press the "Options" button
2. Press the "Push" button

## Expected behavior

The back button should be visible

## Actual Behavior

The back button is not visible

## Your Environment

- React Native Navigation version: 7.8.1
- React Native version: 0.63
- Platform(s) (iOS, Android, or both?): Android
- Device info (Simulator/Device? OS version? Debug/Release?): Simulator / Debug

## Reproducible Demo

https://user-images.githubusercontent.com/5070712/106069526-f5681500-6102-11eb-835d-072e2fa118d3.mp4

### PR DESCRIPTION
**Pull request by yogevbd**
## New Additions
* Add `topBar.backButton.popStackOnPress` which prevents screen pop on default back button press (`true` by default)
* Dispatch `navigationButtonPressed` event on back button press when `popStackOnPress` sets to false
* Fix passing `topBar.backButton.id` option resulted in bad icon resolving.
* Add `hardWareBackButton.dismissModalOnPress` - Controls whether the hardware back button should dismiss modal or not (`true` by default)
* Add `hardWareBackButton.popStackOnPress ` - Controls whether the hardware back button should pop the Stack or not (`true` by default)

## Breaking Changes
until now, passing `topBar.backButton.id` prevented the back button press action from popping the screen. From now on, canceling the popping action is done through `popScreenOnPress` only.

## TODO
- [x] Update [documentation](https://wix.github.io/react-native-navigation/docs/stack#disabling-back-navigation)

Closes #6935

### CODE REVIEW

**Comment by Andarius** [comment]
Is it expected that it won't pop if the **id** field is specified ?

**Comment by yogevbd** [comment]
@Andarius Yes it is expected.

**Comment by guyca** [comment]
@yogevbd I believe `dismissModalOnPress` also needs to be implemented.

File: lib/src/interfaces/Options.ts [hunk 1]
```diff
@@ -387,13 +387,21 @@ export interface OptionsTopBarBackButton {
    */
   displayMode?: 'default' | 'generic' | 'minimal';
   /**
-   * Enables iOS 14 back button menu display
-   * #### (iOS specific)
+   * Controls whether the default back button should pop screen or not
```
Comment (guyca) [inline lib/src/interfaces/Options.ts:390]:
```suggestion
   * Controls whether the default back button should pop the Stack or not
```

**Review by guyca — CHANGES_REQUESTED** [review]

File: e2e/StaticLifecycleEvents.test.js [hunk 2]
```diff
@@ -83,4 +85,49 @@ describe('static lifecycle events', () => {
       elementByLabel('componentDidAppear | CustomRoundedButton | TopBarButton')
     ).toBeVisible();
   })
+
+  it('disabled back button dispatch event and prevents pop', async () => {
```
Comment (guyca) [inline e2e/StaticLifecycleEvents.test.js:89]: The title of the test is a bit misleading as the back button isn't disabled. Moreover, these don't check component lifecycle - I think they belong some place else.

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackController.java [hunk 3]
```diff
@@ -398,12 +403,11 @@ private void startChildrenBellowTopChild() {
         }
     }
 
-    private void onNavigationButtonPressed(String buttonId) {
-        if (Constants.BACK_BUTTON_ID.equals(buttonId)) {
+    private void onNavigationButtonPressed(ButtonOptions button) {
+        if (button.shouldPopOnPress())
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackController.java:407]: @yogevbd Shouldn't we check if the button is a back button? `button.isBackButton()` ? Previously we were checking `if (Constants.BACK_BUTTON_ID.equals(buttonId))`

Comment (guyca) [reply lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackController.java:407]: Ohhh you pulled the comparison into the `shouldPopOnPress` - pretty confusing IMO

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackPresenter.java [hunk 4]
```diff
@@ -541,6 +541,10 @@ private LayoutParams getComponentLayoutParams(ComponentOptions component) {
         return new Toolbar.LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT, component.alignment == Alignment.Center ? Gravity.CENTER : Gravity.START);
     }
 
+    public boolean hardwareBackButtonShouldPop(ViewController viewController) {
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/StackPresenter.java:544]: Typically in functions that return a boolean, the function should begin with the word `should`. Currently, this is sounds yoda speak/french 😅

File: lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/topbar/button/IconResolver.java [hunk 5]
```diff
@@ -36,7 +36,7 @@ public void onError(Throwable error) {
                     throw new RuntimeException(error);
                 }
             });
-        } else if (Constants.BACK_BUTTON_ID.equals(button.id)) {
+        } else if (button.isBackButton()) {
```
Comment (guyca) [inline lib/android/app/src/main/java/com/reactnativenavigation/viewcontrollers/stack/topbar/button/IconResolver.java:39]: 👍

File: lib/src/interfaces/Options.ts [hunk 6]
```diff
@@ -387,6 +386,26 @@ export interface OptionsTopBarBackButton {
...
+export interface HardwareBackButtonOptions {
+  /**
+   * Controls whether the hardware back button should dismiss modal or not
+   * #### (Android specific)
+   * @default true
+   */
+  dismissModalOnPress?: boolean;
+  /**
+   * Controls whether the hardware back button should pop the Stack or not
+   * #### (Android specific)
```
Comment (guyca) [inline lib/src/interfaces/Options.ts:405]: popStackOnPress is relevant for both platforms

File: playground/src/screens/StaticEventsScreen.tsx [hunk 7]
```diff
@@ -26,6 +28,16 @@ export default class StaticEventsScreen extends React.Component<NavigationCompon
           onPress={this.showEventsOverlay}
         />
         <Button label="Push" testID={PUSH_BTN} onPress={this.push} />
+        <Button
```
Comment (guyca) [inline playground/src/screens/StaticEventsScreen.tsx:31]: Sorry for nitpicking for testing but why is this logic in `StaticEventsScreen`? Should we test for back button in Modal screen and in Pushed screen?

**Review by guyca — CHANGES_REQUESTED** [review]

File: lib/src/interfaces/Options.ts (hunk 6 above)
Comment (yogevbd) [reply lib/src/interfaces/Options.ts:405]: But not for the hardware button

Comment (guyca) [reply lib/src/interfaces/Options.ts:405]: w00ps my bad

**Review by guyca — APPROVED** [review]

**Pull request merged**